Fix error when sending build failure email

robolson opened the pull request:
Do not attempt to link to stdout log in failure email if the stdout log file has not been uploaded yet.

moshez commented on app/views/build_mailer/build_break_email.html.haml:
```diff
@@ -20,9 +20,11 @@
-          failed after #{ failed_build_part.elapsed_time.to_i/60 } minutes,
-          for details you can go directly to the #{ link_to('stdout', failed_build_stdout(failed_build_part)) }
+          = link_to("Part number #{ failed_build_part.id }", repository_build_part_url(@build.repository, @build, failed_build_part))
+          failed after #{ failed_build_part.elapsed_time.to_i/60 } minutes
```
This causes an annoying newline between "minutes" and "," which translates to whitespace in the HTML rendering:

I hacked up a unit test to show this, here is the output:

   ...failed after 0 minutes\n,...

Here is the unit test:

    diff --git a/spec/mailers/build_mailer_spec.rb b/spec/mailers/build_mailer_spec.rb
    index 19d6d5e..3378dcc 100644
    --- a/spec/mailers/build_mailer_spec.rb
    +++ b/spec/mailers/build_mailer_spec.rb
    @@ -124,6 +124,7 @@ describe BuildMailer do
               email = BuildMailer.build_break_email(build)
               expect(email.text_part.body).to include(build_artifact_url(stdout_artifact))
               expect(email.html_part.body).to include(build_artifact_url(stdout_artifact))
    +          expect(email.html_part.body).to include('minutes,')
             end
           end
     
Not a big deal, but it would be nice to avoid this visual regression.

robolson commented:
Several additional changes:

1. I moved the weird Build Artifact logic to a decorator like I wanted to.
2. I removed some dead code
3. I fixed the pesky comma issue by moving the code into a Rails helper.

moshez commented on app/decorators/build_part_decorator.rb:
```diff
@@ -0,0 +1,12 @@
+        'build_attempts.build_part_id' => object.id,
+        'build_attempts.state' => BuildAttempt::COMPLETED_BUILD_STATES
+      ).stdout_log.last
+  end
```
Thanks! That looks much better than the ad-hocness previously.

moshez commented:
LGTM
